File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

document.addEventListener("turbo:load", () => {
  const comments = document.getElementById('comments');
  if (comments === null) {
    return;
  }
  const currentUserId = document.body.dataset.currentUserId;
  const roomId = comments.dataset.roomId;
  const appRoom = consumer.subscriptions.create({channel: "RoomChannel", room_id: roomId}, {
    received(data) {
      let commentHtml;
      if (data.sender_id.toString() === currentUserId) {
        commentHtml = data.comment;
      } else {
        commentHtml = data.second_comment;
      }
      
      comments.insertAdjacentHTML('beforeend', commentHtml);
    },

    speak: function(comment, roomId) {
      return this.perform('speak', {comment: comment, room_id: roomId});
    }
  });

  const input = document.getElementById('room_input');
  const button = document.getElementById('submit_button');

  if(input) {
    input.addEventListener("keypress", function(e) {
      if (e.key === 'Enter') {
        appRoom.speak(e.target.value, roomId);
        e.target.value = '';
        e.preventDefault();
      }
    });
  } else {
    console.error("Could not find element with id 'room_input'");
  }

  if(button) {
    button.addEventListener("click", function(e) {
      if(input.value !== '') {
        appRoom.speak(input.value, roomId);
        input.value = '';
      }
      e.preventDefault();
    });
  } else {
    console.error("Could not find element with id 'submit_button'");
  }
});

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    room = Room.find(params[:room_id])
    stream_for room
  end

  def unsubscribed; end

  def speak(data)
    room = Room.find(data['room_id'])
    body = data['comment'].strip # メッセージの前後の空白を削除
    return if body.blank? # メッセージが空でないことを確認
    # メッセージの作成だけ行い、ブロードキャストは after_commit フックに任せる
    room.comments.create!(body:, user_id: current_user.id)
    sleep(0.1)
  end
end
